transcom/mymove
# INT- B -18876- Prime ListMove amendment count

## [B-18876](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A907603)

## Summary

**INTEGRATION**

This PR updates Prime ListMoves API (prime.yaml).  

`Prime V1  - GET: {{baseUrl}}/moves`

Query string parameter:
- since (optional)
  - example:` {{baseUrl}}/moves?since=2024-04-12T17:35:06.728Z`

Note: Updating existing Prime v1 of listMoves API. This enhancement is just adding NEW field to the ListMove schema. We are not breaking a contract between producer and consumer.

The update adds "amendments" to the ListMove schema. 
![Screenshot 2024-04-12 at 1 47 17 PM](https://github.com/transcom/mymove/assets/148127004/f401a058-67ef-466c-9a93-60a4c2497ac5)

**total:** represents total amendments for move 
**availableSince:** represents the number of amendments uploaded per "since" parameter.  It's basically the delta of number of amendments uploaded since timestamp.


## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Log in as a customer. Create a move with HHG. The easiest way to test is to use a specific pickup zip code(IL, 62226) and destination zip code(MA, 02542) to ensure the same GBLOC will be used for both pickup and destination.   This will result in using AGFM for both New Services Counseling and TOO logins.
2. Log into New Services Counseling and `<Submit move details> `for move
3. Log into TOO and `<Approve selected>` for move. Note: move code will not be visible on the listMoves unless task order is approved.
4. Using Swagger or Postman run test case:
     - 1: default -  no "since"  query string param filter
        - This will return everything in the system. If move has amendments total and available counts will be the same.
     - 2: apply since filter
        -  This will return the delta count if amendments have been uploaded.  delta:  x >= since  

Sql verification (apologizes for the format of the sql)  Modify the timestamp and locator code and run.
`
WITH mainSubQuery AS (
	SELECT moves.*
	FROM   moves
		   inner join orders
				   ON moves.orders_id = orders.id
	WHERE  moves.available_to_prime_at IS NOT NULL
		   AND moves.show = TRUE
		   AND ( moves.updated_at >= '2024-04-12T17:35:06.728Z'
				  OR orders.updated_at >= '2024-04-12T17:35:06.728Z'
				  OR ( moves.id IN (SELECT mto_shipments.move_id
									FROM   mto_shipments
									WHERE  mto_shipments.updated_at >=
										   '2024-04-12T17:35:06.728Z'
									UNION
									SELECT mto_service_items.move_id
									FROM   mto_service_items
									WHERE  mto_service_items.updated_at >=
										   '2024-04-12T17:35:06.728Z'
									UNION
									SELECT payment_requests.move_id
									FROM   payment_requests
									WHERE  payment_requests.updated_at >=
										   '2024-04-12T17:35:06.728Z'
									UNION
									SELECT mto_shipments.move_id
									FROM   mto_shipments
										   inner join reweighs
												   ON reweighs.shipment_id =
													  mto_shipments.id
									WHERE  reweighs.updated_at >=
										   '2024-04-12T17:35:06.728Z') ) )
),
ordersWithAmendmentsSubQuery AS (
    select orders.id orders_id, documents.id document_id, user_uploads.created_at 
    from orders join documents on orders.uploaded_amended_orders_id = documents.id
    join user_uploads on user_uploads.document_id = documents.id
),
ordersWithAmendmentsTotalCountSubQuery AS (
    select orders_id, document_id, count(document_id) amendments_total from ordersWithAmendmentsSubQuery group by document_id, orders_id
),
ordersWithAmendmentsSinceTotalCountSubQuery AS (
    select since_count_sub.orders_id, since_count_sub.document_id, count(since_count_sub.document_id) amendments_since_total from (
        select * from ordersWithAmendmentsSubQuery where created_at >= '2024-04-12T17:35:06.728Z'
    ) since_count_sub  group by since_count_sub.document_id, since_count_sub.orders_id
),
combineTotalCountSubQueries AS (
    select ordersWithAmendmentsTotalCountSubQuery.orders_id, ordersWithAmendmentsTotalCountSubQuery.amendments_total,  
    ordersWithAmendmentsSinceTotalCountSubQuery.amendments_since_total from ordersWithAmendmentsTotalCountSubQuery 
    left outer join ordersWithAmendmentsSinceTotalCountSubQuery 
    on ordersWithAmendmentsTotalCountSubQuery.orders_id = ordersWithAmendmentsSinceTotalCountSubQuery.orders_id
)
select  mainSubQuery.locator, combineTotalCountSubQueries.* from mainSubQuery left outer join combineTotalCountSubQueries on mainSubQuery.orders_id = combineTotalCountSubQueries.orders_id
where mainSubQuery.locator = 'R443MB';
`


### Backend
